File: concept/value/ValueImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.value;

import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.Concept;
import com.vaticle.typedb.client.api.concept.type.Type;
import com.vaticle.typedb.client.api.concept.type.Type.ValueType;
import com.vaticle.typedb.client.api.concept.value.Value;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.concept.ConceptImpl;
import com.vaticle.typedb.protocol.ConceptProto;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.BAD_VALUE_TYPE;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.INVALID_CONCEPT_CASTING;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.VALUE_HAS_NO_REMOTE;
import static com.vaticle.typedb.common.util.Objects.className;

public abstract class ValueImpl<VALUE> extends ConceptImpl implements Value<VALUE> {

    protected final VALUE value;
    private final ValueType valueType;

    public ValueImpl(ValueType valueType, VALUE value) {
        this.valueType = valueType;
        this.value = value;
    }

    public static ValueImpl<?> of(ConceptProto.Value valueProto) {
        switch (valueProto.getValueType()) {
            case BOOLEAN:
                return ValueImpl.Boolean.of(valueProto);
            case LONG:
                return ValueImpl.Long.of(valueProto);
            case DOUBLE:
                return ValueImpl.Double.of(valueProto);
            case STRING:
                return ValueImpl.String.of(valueProto);
            case DATETIME:
                return ValueImpl.DateTime.of(valueProto);
            case UNRECOGNIZED:
            default:
                throw new TypeDBClientException(BAD_VALUE_TYPE, valueProto.getValueType());
        }
    }

    @Override
    public ValueImpl<?> asValue() {
        return this;
    }

    @Override
    public Concept.Remote asRemote(TypeDBTransaction transaction) {
        throw new TypeDBClientException(VALUE_HAS_NO_REMOTE);
    }

    @Override
    public ValueType getValueType() {
        return valueType;
    }

    @Override
    public VALUE getValue() {
        return this.value;
    }

    @Override
    public Value<java.lang.Boolean> asBoolean() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Value.Boolean.class));
    }

    @Override
    public Value<java.lang.Long> asLong() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Value.Long.class));
    }

    @Override
    public Value<java.lang.Double> asDouble() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Value.Double.class));
    }

    @Override
    public Value<java.lang.String> asString() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Value.String.class));
    }

    @Override
    public Value<LocalDateTime> asDateTime() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Value.DateTime.class));
    }

    public static class Boolean extends ValueImpl<java.lang.Boolean> implements Value.Boolean {
        public Boolean(ValueType valueType, java.lang.Boolean value) {
            super(valueType, value);
        }

        public static ValueImpl.Boolean of(ConceptProto.Value valueProto) {
            assert valueProto.getValueType() == ConceptProto.ValueType.BOOLEAN;
            return new ValueImpl.Boolean(ValueType.of(valueProto.getValueType()), valueProto.getValue().getBoolean());
        }

        public ValueImpl.Boolean asBoolean() {
            return this;
        }
    }

    public static class Long extends ValueImpl<java.lang.Long> implements Value.Long {
        public Long(ValueType valueType, java.lang.Long value) {
            super(valueType, value);
        }

        public static ValueImpl.Long of(ConceptProto.Value valueProto) {
            assert valueProto.getValueType() == ConceptProto.ValueType.LONG;
            return new ValueImpl.Long(ValueType.of(valueProto.getValueType()), valueProto.getValue().getLong());
        }

        public ValueImpl.Long asLong() {
            return this;
        }
    }

    public static class Double extends ValueImpl<java.lang.Double>  implements Value.Double {
        public Double(ValueType valueType, java.lang.Double value) {
            super(valueType, value);
        }

        public static ValueImpl.Double of(ConceptProto.Value valueProto) {
            assert valueProto.getValueType() == ConceptProto.ValueType.DOUBLE;
            return new ValueImpl.Double(ValueType.of(valueProto.getValueType()), valueProto.getValue().getDouble());
        }

        public ValueImpl.Double asDouble() {
            return this;
        }
    }

    public static class String extends ValueImpl<java.lang.String> implements Value.String {
        public String(ValueType valueType, java.lang.String value) {
            super(valueType, value);
        }

        public static ValueImpl.String of(ConceptProto.Value valueProto) {
            assert valueProto.getValueType() == ConceptProto.ValueType.STRING;
            return new ValueImpl.String(ValueType.of(valueProto.getValueType()), valueProto.getValue().getString());
        }

        public ValueImpl.String asString() {
            return this;
        }
    }

    public static class DateTime extends ValueImpl<LocalDateTime> implements Value.DateTime {
        public DateTime(ValueType valueType, LocalDateTime value) {
            super(valueType, value);
        }

        public static ValueImpl.DateTime of(ConceptProto.Value valueProto) {
            assert valueProto.getValueType() == ConceptProto.ValueType.DATETIME;
            return new ValueImpl.DateTime(ValueType.of(valueProto.getValueType()),
                    LocalDateTime.ofInstant(Instant.ofEpochMilli(valueProto.getValue().getDateTime()), ZoneOffset.UTC));
        }

        public ValueImpl.DateTime asDateTime() {
            return this;
        }
    }
}

File: concept/value/ValueImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {Value} from "../../api/concept/value/Value";
import {ConceptImpl} from "../ConceptImpl";
import {Concept} from "../../api/concept/Concept";
import {TypeDBTransaction} from "../../api/connection/TypeDBTransaction";
import {TypeDBClientError} from "../../common/errors/TypeDBClientError";
import {ErrorMessage} from "../../common/errors/ErrorMessage";
import {Value as ValueProto, ValueType as ValueTypeProto} from "typedb-protocol/common/concept_pb";
import ValueType = Concept.ValueType;
import VALUE_HAS_NO_REMOTE = ErrorMessage.Concept.VALUE_HAS_NO_REMOTE;
import INVALID_CONCEPT_CASTING = ErrorMessage.Concept.INVALID_CONCEPT_CASTING;

import BAD_VALUE_TYPE = ErrorMessage.Concept.BAD_VALUE_TYPE;

export abstract class ValueImpl extends ConceptImpl implements Value {

    private readonly _valueType: ValueType;

    protected constructor(type: ValueType) {
        super();
        this._valueType = type;
    }

    get valueType(): ValueType {
        return this._valueType;
    }

    abstract get value(): boolean | string | number | Date;

    asRemote(transaction: TypeDBTransaction): Concept.Remote {
        throw new TypeDBClientError(VALUE_HAS_NO_REMOTE);
    }

    isValue(): boolean {
        return true
    }

    asValue(): Value {
        return this;
    }

    equals(concept: Concept): boolean {
        if (!concept.isValue()) return false;
        else {
            return this.valueType == concept.asValue().valueType && this.value == concept.asValue().value;
        }
    }

    toJSONRecord(): Record<string, boolean | string | number> {
        let value;
        if (this.value instanceof Date) value = this.value.toISOString().slice(0, -1);
        else value = this.value;
        return {
            value_type: this.valueType.name(),
            value: value
        };
    }

    isBoolean(): boolean {
        return false;
    }

    isDateTime(): boolean {
        return false;
    }

    isDouble(): boolean {
        return false;
    }

    isLong(): boolean {
        return false;
    }

    isString(): boolean {
        return false;
    }

    asBoolean(): Value.Boolean {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Value.Boolean"));
    }

    asLong(): Value.Long {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Value.Long"));
    }

    asDouble(): Value.Double {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Value.Double"));
    }

    asString(): Value.String {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Value.String"));
    }

    asDateTime(): Value.DateTime {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Value.DateTime"));
    }
}

export namespace ValueImpl {

    export function of(valueProto: ValueProto): Value {
        if (!valueProto) return null;
        switch (valueProto.getValueType()) {
            case ValueTypeProto.BOOLEAN:
                return new ValueImpl.Boolean(valueProto.getValue().getBoolean());
            case ValueTypeProto.LONG:
                return new ValueImpl.Long(valueProto.getValue().getLong());
            case ValueTypeProto.DOUBLE:
                return new ValueImpl.Double(valueProto.getValue().getDouble());
            case ValueTypeProto.STRING:
                return new ValueImpl.String(valueProto.getValue().getString());
            case ValueTypeProto.DATETIME:
                return new ValueImpl.DateTime(new Date(valueProto.getValue().getDateTime()));
            default:
                throw new TypeDBClientError(BAD_VALUE_TYPE.message(valueProto.getValueType()));
        }
    }

    export class Boolean extends ValueImpl implements Value.Boolean {
        readonly _value: boolean;

        constructor(value: boolean) {
            super(ValueType.BOOLEAN);
            this._value = value;
        }

        get value(): boolean {
            return this._value;
        }

        isBoolean(): boolean {
            return true;
        }

        asBoolean(): Value.Boolean {
            return this;
        }

        protected get className(): string {
            return "Value.Boolean";
        }
    }

    export class Long extends ValueImpl implements Value.Long {
        readonly _value: number;

        constructor(value: number) {
            super(ValueType.LONG);
            this._value = value;
        }

        get value(): number {
            return this._value;
        }

        isLong(): boolean {
            return true;
        }

        asLong(): Value.Long {
            return this;
        }

        protected get className(): string {
            return "Value.Long";
        }
    }

    export class Double extends ValueImpl implements Value.Double {
        readonly _value: number;

        constructor(value: number) {
            super(ValueType.DOUBLE);
            this._value = value;
        }

        get value(): number {
            return this._value;
        }

        isDouble(): boolean {
            return true;
        }

        asDouble(): Value.Double {
            return this;
        }

        protected get className(): string {
            return "Value.Double";
        }
    }

    export class String extends ValueImpl implements Value.String {
        readonly _value: string;

        constructor(value: string) {
            super(ValueType.STRING);
            this._value = value;
        }

        get value(): string {
            return this._value;
        }

        isString(): boolean {
            return true;
        }

        asString(): Value.String {
            return this;
        }

        protected get className(): string {
            return "Value.String";
        }
    }

    export class DateTime extends ValueImpl implements Value.DateTime {
        readonly _value: Date;

        constructor(value: Date) {
            super(ValueType.DATETIME);
            this._value = value;
        }

        get value(): Date {
            return this._value;
        }

        isDateTime(): boolean {
            return true;
        }

        asDateTime(): Value.DateTime {
            return this;
        }

        protected get className(): string {
            return "Value.DateTime";
        }
    }

}